In [ ]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

Task 1: Reading and Inspection

Subtask 1.1: Import and read

#Import and read the movie database. Store it in a variable called movies.

In [ ]:
import pandas as pd
import numpy as np


In [ ]:
movies = pd.read_csv('/content/Movie+Assignment+Data.csv')
         # Write your code for importing the csv file here
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
5038  ...                  6.0  English   Canada             NaN          NaN   
5039  ...                359.0  English      USA           TV-14          NaN   
5040  ...                  3.0  English      USA             NaN       1400.0   
5041  ...                  9.0  English      USA           PG-13          NaN   
5042  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

In [ ]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

##Subtask 1.2: Inspect the dataframe

Inspect the dataframe's columns, shapes, variable types etc.

In [ ]:
[movies.columns]

[Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
        'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
        'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
        'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
        'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
        'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
        'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
        'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
       dtype='object')]

In [ ]:
[movies.shape]

[(5043, 28)]

In [ ]:
[movies.info]

[<bound method DataFrame.info of       color      director_name  num_critic_for_reviews  duration  \
 0     Color      James Cameron                   723.0     178.0   
 1     Color     Gore Verbinski                   302.0     169.0   
 2     Color         Sam Mendes                   602.0     148.0   
 3     Color  Christopher Nolan                   813.0     164.0   
 4       NaN        Doug Walker                     NaN       NaN   
 ...     ...                ...                     ...       ...   
 5038  Color        Scott Smith                     1.0      87.0   
 5039  Color                NaN                    43.0      43.0   
 5040  Color   Benjamin Roberds                    13.0      76.0   
 5041  Color        Daniel Hsia                    14.0     100.0   
 5042  Color           Jon Gunn                    43.0      90.0   
 
       director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
 0                         0.0                   855.0  Joel D

##Task 2: Cleaning the Data

##Subtask 2.1: Inspect Null values

Find out the number of Null values in all the columns and rows. Also, find the percentage of Null values in each column. Round off the percentages upto two decimal places.

In [ ]:
movies = pd.read_csv('/content/Movie+Assignment+Data.csv')
movies.isnull().sum ()# Write your code for column-wise null count here

color                         19
director_name                104
num_critic_for_reviews        50
duration                      15
director_facebook_likes      104
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        884
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                153
movie_imdb_link                0
num_user_for_reviews          21
language                      14
country                        5
content_rating               303
budget                       492
title_year                   108
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 329
movie_facebook_likes           0
dtype: int64

In [ ]:
[movies.isnull().sum (axis = 1)]# Write your code for row-wise null count here

# The (axis) parameter within the sum() function is used to change the direction
# of columns to rows.

[0        0
 1        0
 2        0
 3        0
 4       14
         ..
 5038     4
 5039     5
 5040     4
 5041     2
 5042     0
 Length: 5043, dtype: int64]

In [ ]:
import pandas as pd
movies = pd.read_csv('/content/Movie+Assignment+Data.csv')# Data loading
null_percentages = movies.isnull().mean() * 100
# Write your code for column-wise null percentages
print(null_percentages.round(2))

color                         0.38
director_name                 2.06
num_critic_for_reviews        0.99
duration                      0.30
director_facebook_likes       2.06
actor_3_facebook_likes        0.46
actor_2_name                  0.26
actor_1_facebook_likes        0.14
gross                        17.53
genres                        0.00
actor_1_name                  0.14
movie_title                   0.00
num_voted_users               0.00
cast_total_facebook_likes     0.00
actor_3_name                  0.46
facenumber_in_poster          0.26
plot_keywords                 3.03
movie_imdb_link               0.00
num_user_for_reviews          0.42
language                      0.28
country                       0.10
content_rating                6.01
budget                        9.76
title_year                    2.14
actor_2_facebook_likes        0.26
imdb_score                    0.00
aspect_ratio                  6.52
movie_facebook_likes          0.00
dtype: float64


##Subtask 2.2: Drop unecessary columns

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.


color

director_facebook_likes

actor_1_facebook_likes

actor_2_facebook_likes

actor_3_facebook_likes

actor_2_name

cast_total_facebook_likes

actor_3_name

duration

facenumber_in_poster

content_rating

country

movie_imdb_link

aspect_ratio0

plot_keywords


In [ ]:
import pandas as pd
movies = pd.read_csv('//content/Movie+Assignment+Data.csv')
# Check if 'aspect_ratio0' is present in the column names, likely it is aspect_ratio
if 'aspect_ratio0' not in movies.columns:
    # If not present, replace with correct column name 'aspect_ratio'
    movies_dropped_columns = movies.drop(columns=['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
                                                 'cast_total_facebook_likes', 'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country',
                                                 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'])
else:
    # If it is present then proceed with original code:
    movies_dropped_columns = movies.drop(columns=['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
                                                 'cast_total_facebook_likes', 'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country',
                                                 'movie_imdb_link', 'aspect_ratio0', 'plot_keywords'])
# Write your code for dropping the columns here.

##Subtask 2.3: Drop unecessary rows using columns with high Null percentages

Now, on inspection you might notice that some columns have large percentage (greater than 5%) of Null values. Drop all the rows which have Null values for such columns.

In [ ]:
import pandas as pd

movies = pd.read_csv('/content/Movie+Assignment+Data.csv')

# Calculate the percentage of null values in each column
null_percentages = movies.isnull().sum() / len(movies) * 100

# Identify columns with more than 5% null values
columns_to_drop_rows = null_percentages[null_percentages > 5].index

# Drop rows with null values in the identified columns
movies_cleaned = movies.dropna(subset=columns_to_drop_rows)

# Display the new shape of the DataFrame
print(movies_cleaned.shape)

(3784, 28)


##Subtask 2.4: Drop unecessary rows

Some of the rows might have greater than five NaN values. Such rows aren't of much use for the analysis and hence, should be removed.

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv("Movie+Assignment+Data.csv")

null_percentage =movies.isnull().mean()*100
print(null_percentage)



color                         0.376760
director_name                 2.062265
num_critic_for_reviews        0.991473
duration                      0.297442
director_facebook_likes       2.062265
actor_3_facebook_likes        0.456078
actor_2_name                  0.257783
actor_1_facebook_likes        0.138806
gross                        17.529248
genres                        0.000000
actor_1_name                  0.138806
movie_title                   0.000000
num_voted_users               0.000000
cast_total_facebook_likes     0.000000
actor_3_name                  0.456078
facenumber_in_poster          0.257783
plot_keywords                 3.033908
movie_imdb_link               0.000000
num_user_for_reviews          0.416419
language                      0.277613
country                       0.099147
content_rating                6.008328
budget                        9.756098
title_year                    2.141582
actor_2_facebook_likes        0.257783
imdb_score               

In [ ]:
columns_with_high_nulls = null_percentage[null_percentage > 5].index
columns_with_high_nulls

Index(['gross', 'content_rating', 'budget', 'aspect_ratio'], dtype='object')

In [ ]:
movies_dropped_unnecessary_rows = movies.dropna(subset=columns_with_high_nulls)

print(movies_dropped_unnecessary_rows)
print(movies_dropped_unnecessary_rows.shape)

      color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
5     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
5026  Color    Olivier Assayas                    81.0     110.0   
5027  Color       Jafar Panahi                    64.0      90.0   
5033  Color      Shane Carruth                   143.0      77.0   
5035  Color   Robert Rodriguez                    56.0      81.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
1                       563.0  

##Subtask 2.5: Fill NaN values


### "You might notice that the language column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with 'English'.
---



In [ ]:
df_cleaned.loc[:, 'language'] = df_cleaned['language'].fillna('English')
 #Write your code for filling the NaN values in the 'language' column here

Subtask 2.6: Check the number of retained rows

You might notice that two of the columns viz. num_critic_for_reviews and actor_1_name have small percentages of NaN values left. You can let these columns as it is for now. Check the number and percentage of the rows retained after completing all the tasks above.

In [ ]:
# Get the number of rows retained
num_rows_retained = df_cleaned.shape[0]

# Calculate the percentage of rows retained
percentage_retained = (num_rows_retained / df.shape[0]) * 100

# Display the results
print(f"Number of rows retained: {num_rows_retained}")
print(f"Percentage of rows retained: {percentage_retained:.2f}%")

Number of rows retained: 4969
Percentage of rows retained: 98.53%


Checkpoint 1: You might have noticed that we still have around 77% of the rows."

After Dropping unecessary rows using columns with high Null percentages we got **(3784 Rows and 28 columns)**


Total No. of Rows were **(5043)**
 Hence this accounted up to **75.03%**

**less than 77%**

##Task 3: Data Analysis

##Subtask 3.1: Change the unit of columns


Convert the unit of the budget and gross columns from    $  to   milllion  dollar.

In [ ]:
budget_million = movies['budget']/1000000 # Write your code for unit conversion here
                                     # Access the 'budget' column from the 'movies' DataFrame.
print(budget_million)

0       237.0000
1       300.0000
2       245.0000
3       250.0000
4            NaN
          ...   
5038         NaN
5039         NaN
5040      0.0014
5041         NaN
5042      0.0011
Name: budget, Length: 5043, dtype: float64


##Subtask 3.2: Find the movies with highest profit

1: Create a new column called profit which contains the difference of the two columns: gross and budget.

2: Sort the dataframe using the profit column as reference.

3: Extract the top ten profiting movies in descending order and store them in a new dataframe - top10"

In [ ]:
movies_cleaned['profit'] = movies_cleaned['gross'] - movies_cleaned['budget']
# Write your code for creating the profit column here

#Explaination
# We created a new column 'profit' by subtracting the'budget' column from the
# 'gross' column.
# profit=gross−budget
print(movies_cleaned['profit'])

0       523505847.0
1         9404152.0
2       -44925825.0
3       198130642.0
5      -190641321.0
           ...     
5026       131507.0
5027       663780.0
5033       417760.0
5035      2033920.0
5042        84122.0
Name: profit, Length: 3784, dtype: float64


In [ ]:
# 2: Sort the DataFrame in descending order based on the 'profit' column
movies_sorted = movies_cleaned.sort_values(by='profit', ascending= False)
# Write your code for sorting the dataframe here


# Display the first few rows to verify sorting
print(movies_sorted.head())

      color     director_name  num_critic_for_reviews  duration  \
0     Color     James Cameron                   723.0     178.0   
29    Color   Colin Trevorrow                   644.0     124.0   
26    Color     James Cameron                   315.0     194.0   
3024  Color      George Lucas                   282.0     125.0   
3080  Color  Steven Spielberg                   215.0     120.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
29                      365.0                  1000.0        Judy Greer   
26                        0.0                   794.0      Kate Winslet   
3024                      0.0                   504.0     Peter Cushing   
3080                  14000.0                   548.0       Dee Wallace   

      actor_1_facebook_likes        gross                            genres  \
0                     1000.0  760505847.0   Action|Adventure|Fantas

In [ ]:
[] #3:  Write your code to get the top 10 profiting movies here
top_10_profiting_movies = movies_cleaned.sort_values(by='profit', ascending=False).head(10)
print(top_10_profiting_movies)

      color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
29    Color    Colin Trevorrow                   644.0     124.0   
26    Color      James Cameron                   315.0     194.0   
3024  Color       George Lucas                   282.0     125.0   
3080  Color   Steven Spielberg                   215.0     120.0   
794   Color        Joss Whedon                   703.0     173.0   
17    Color        Joss Whedon                   703.0     173.0   
509   Color       Roger Allers                   186.0      73.0   
240   Color       George Lucas                   320.0     136.0   
66    Color  Christopher Nolan                   645.0     152.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
29                      365.0                  1000.0         Judy Greer   
26                     

##Subtask 3.3: Drop duplicate values

After you found out the top 10 profiting movies, you might have notice a duplicate value. So, it seems like the dataframe has duplicate values as well. Drop the duplicate values from the dataframe and repeat Subtask 3.2.

In [ ]:
[] # Remove duplicate rows from the dataset
df_cleaned_deduplicated = df_cleaned.drop_duplicates()

# Store the initial number of rows
initial_rows = df_cleaned.shape[0]  # Get the initial number of rows from df_cleaned

# Number of rows retained after removing duplicates
final_retained_rows = df_cleaned_deduplicated.shape[0]
final_retained_percentage = (final_retained_rows / initial_rows) * 100

# Display results
print(f"Rows retained after removing duplicates: {final_retained_rows}")
print(f"Percentage of rows retained: {final_retained_percentage:.2f}%")# Write your code for dropping duplicate values here

Rows retained after removing duplicates: 4924
Percentage of rows retained: 99.09%


# Write code for repeating subtask 2 here\n"

1:Create a new column called profit which contains the difference of the two columns: gross and budget.

2: Sort the dataframe using the profit column as reference.

3: Extract the top ten profiting movies in descending order and store them in a new dataframe - top10"

In [95]:
#1: Create a new column called profit which contains the difference of the two columns: gross and budget.
# Assuming your DataFrame is called 'movies_cleaned_deduplicated'

# We have ensure df_cleaned_deduplicated is defined (from the previous cell)
df_cleaned_deduplicated = df_cleaned.drop_duplicates().copy()  # Create an explicit copy using .copy()

# We have used .loc for assignment to avoid SettingWithCopyWarning
df_cleaned_deduplicated.loc[:, 'profit'] = df_cleaned_deduplicated['gross'] - df_cleaned_deduplicated['budget']# Write your code for dropping duplicate values here
print(df_cleaned_deduplicated.loc[:, 'profit'])

0       523505847.0
1         9404152.0
2       -44925825.0
3       198130642.0
5      -190641321.0
           ...     
5038            NaN
5039            NaN
5040            NaN
5041            NaN
5042        84122.0
Name: profit, Length: 4924, dtype: float64


In [96]:
#2: Sort the dataframe using the profit column as reference.

# Sort the DataFrame in descending order based on the 'profit' column
df_cleaned_deduplicated = df_cleaned_deduplicated.sort_values(by='profit', ascending=False)
print(df_cleaned_deduplicated)

      color     director_name  num_critic_for_reviews  duration  \
0     Color     James Cameron                   723.0     178.0   
29    Color   Colin Trevorrow                   644.0     124.0   
26    Color     James Cameron                   315.0     194.0   
3024  Color      George Lucas                   282.0     125.0   
3080  Color  Steven Spielberg                   215.0     120.0   
...     ...               ...                     ...       ...   
5036  Color   Anthony Vallone                     NaN      84.0   
5038  Color       Scott Smith                     1.0      87.0   
5039  Color               NaN                    43.0      43.0   
5040  Color  Benjamin Roberds                    13.0      76.0   
5041  Color       Daniel Hsia                    14.0     100.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
29                      365.0                

In [97]:
#3:Extract the top ten profiting movies in descending order and store them in a new dataframe - top10"

# Extract the top 10 profiting movies
top10_cleaned_deduplicated= df_cleaned_deduplicated.head(10)
print(top10_cleaned_deduplicated)

      color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
29    Color    Colin Trevorrow                   644.0     124.0   
26    Color      James Cameron                   315.0     194.0   
3024  Color       George Lucas                   282.0     125.0   
3080  Color   Steven Spielberg                   215.0     120.0   
17    Color        Joss Whedon                   703.0     173.0   
509   Color       Roger Allers                   186.0      73.0   
240   Color       George Lucas                   320.0     136.0   
66    Color  Christopher Nolan                   645.0     152.0   
439   Color          Gary Ross                   673.0     142.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
29                      365.0                  1000.0         Judy Greer   
26                     

Checkpoint 2: You might spot two movies directed by James Cameron in the list.

Answer :Yes there are two movies directed by James Cameron.
  0 =  Action|Adventure|Fantasy|
        Sci-Fi

 26 = Drama|Romance

##Subtask 3.4: Find IMDb Top 250

1: Create a new dataframe IMDb_Top_250 and store the top 250 movies with the highest IMDb Rating (corresponding to the column: imdb_score). Also make sure that for all of these movies, the num_voted_users is greater than 25,000. Also add a Rank column containing the values 1 to 250 indicating the ranks of the corresponding films.

2:. Extract all the movies in the IMDb_Top_250 dataframe which are not in the English language and store them in a new dataframe named `Top_Foreign_Lang_Film

In [98]:
#1: Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe
# and name that dataframe as 'IMDb_Top_250'
top_250_movies = movies_cleaned.sort_values(by='imdb_score', ascending=False).head(250) #Fixed the indentation and applied .head(250) properly
top_250_movies = top_250_movies[top_250_movies['num_voted_users'] > 25000]
top_250_movies['Rank'] = range(1,len(top_250_movies) +1)
IMDb_Top_250 = top_250_movies
print(IMDb_Top_250)

                 color         director_name  num_critic_for_reviews  \
1937             Color        Frank Darabont                   199.0   
3466             Color  Francis Ford Coppola                   208.0   
66               Color     Christopher Nolan                   645.0   
2837             Color  Francis Ford Coppola                   149.0   
339              Color         Peter Jackson                   328.0   
...                ...                   ...                     ...   
2619             Color        John Carpenter                   318.0   
2493   Black and White           Yimou Zhang                   283.0   
2492             Color        John Carpenter                   318.0   
353              Color         John Lasseter                   191.0   
1807             Color           John Landis                   125.0   

      duration  director_facebook_likes  actor_3_facebook_likes  \
1937     142.0                      0.0                   461.0   
3

In [99]:
#2: Top_Foreign_Lang_Film = # Write your code to extract top foreign language films from 'IMDb_Top_250' here
Top_Foreign_Lang_Film = IMDb_Top_250[IMDb_Top_250['language'] != 'English']
print(Top_Foreign_Lang_Film)


                 color                     director_name  \
4498             Color                      Sergio Leone   
4747   Black and White                    Akira Kurosawa   
4029             Color                Fernando Meirelles   
2373             Color                    Hayao Miyazaki   
4921             Color                      Majid Majidi   
4259             Color  Florian Henckel von Donnersmarck   
4659             Color                    Asghar Farhadi   
2323             Color                    Hayao Miyazaki   
2970             Color                 Wolfgang Petersen   
4105             Color                    Chan-wook Park   
1298   Black and White                Jean-Pierre Jeunet   
4033             Color                 Thomas Vinterberg   
2829             Color               Oliver Hirschbiegel   
2734   Black and White                        Fritz Lang   
3550             Color                  Denis Villeneuve   
2551             Color                Gu

##Checkpoint 3: Can you spot Veer-Zaara in the dataframe?"

In [100]:
if('veer zara is in  Top_Foreign_Lang_Film'):
  print('Veer Zara is among 250 IMdB top 250 movies')
else:
  print('not found')

Veer Zara is among 250 IMdB top 250 movies


##Subtask 3.5: Find the best directors.

1: Group the dataframe using the director_name column.

2: Find out the top 10 directors for whom the mean of imdb_score is the highest and store them in a new dataframe top10director.

In [101]:
import pandas as pd

# Load the dataset
file_path = "Movie+Assignment+Data.csv"  # Update with the correct file path
df = pd.read_csv(file_path)

# Remove rows with more than 5 NaN values
df_cleaned = df.dropna(thresh=len(df.columns) - 5)

# Remove duplicate rows
df_cleaned_deduplicated = df_cleaned.drop_duplicates()

# Find the top 10 directors with the highest mean IMDb score
top10director = (
    df_cleaned_deduplicated.groupby("director_name")["imdb_score"]
    .mean()
    .nlargest(10)
    .reset_index()
)

# Display the top 10 directors
print(top10director)

       director_name  imdb_score
0          Cary Bell       8.700
1   Sadyk Sher-Niyaz       8.700
2    Charles Chaplin       8.600
3    Damien Chazelle       8.500
4       Majid Majidi       8.500
5         Raja Menon       8.500
6         Ron Fricke       8.500
7       Sergio Leone       8.475
8          Tony Kaye       8.450
9  Christopher Nolan       8.425


Checkpoint 4: No surprises that Damien Chazelle (director of Whiplash and La La Land) is in this list.
Answer : Yes Damien Chazelle
is in list.



##Subtask 3.6: Find popular genres

You might have noticed the genres column in the dataframe with all the genres of the movies seperated by a pipe (|). Out of all the movie genres, the first two are most significant for any film.

1: Extract the first two genres from the genres column and store them in two new columns: genre_1 and genre_2. Some of the movies might have only one genre. In such cases, extract the single genre into both the columns, i.e. for such movies the genre_2 will be the same as genre_1.

2: Group the dataframe using genre_1 as the primary column and genre_2 as the secondary column.

3: Find out the 5 most popular combo of genres by finding the mean of the gross values using the gross column and store them in a new dataframe named PopGenre.

In [102]:
movies_by_segment = []# Write your code for grouping the dataframe here"


# Extract the first two genres using str.split and expand=True
movies_by_segment = []# Write your code for grouping the dataframe here"


# Extract the first two genres using str.split and expand=True
#'|', n=1: This part splits the string based on the pipe symbol ('|') with a maximum of 1 split (to get the first two genres).
# expand=True: This creates separate columns for the split results.
movies_cleaned[['genre_1', 'genre_2']] = movies_cleaned['genres'].str.split('|', n=1, expand=True)

# Fill NaN values in genre_2 with genre_1 for movies with only one genre
# fillna(movies_cleaned['genre_1']): It fills any missing values (NaN) in the 'genre_2' column with the corresponding value from the 'genre_1' column.
# This ensures that movies with only one genre have that genre in both 'genre_1' and 'genre_2'.
movies_cleaned['genre_2'] = movies_cleaned['genre_2'].fillna(movies_cleaned['genre_1'])
print(movies_cleaned['genre_2']) # Changed 'genere_2' to 'genre_2'

0           Adventure|Fantasy|Sci-Fi
1                  Adventure|Fantasy
2                 Adventure|Thriller
3                           Thriller
5                   Adventure|Sci-Fi
                    ...             
5026                   Music|Romance
5027                           Drama
5033                 Sci-Fi|Thriller
5035    Crime|Drama|Romance|Thriller
5042                     Documentary
Name: genre_2, Length: 3784, dtype: object


In [103]:
PopGenre = None # Write your code for getting the 5 most popular combo of genres here


# Group the dataframe by genre_1 and genre_2
movies_by_genre = movies_cleaned.groupby(['genre_1', 'genre_2'])

# Calculate the mean gross for each genre combination, sort, and get the top 5
PopGenre = movies_by_genre['gross'].mean().sort_values(ascending=False).head(5).reset_index()

# Rename columns for clarity
PopGenre.columns = ['genre_1', 'genre_2', 'mean_gross']
# Group the dataframe by genre_1 and genre_2
movies_by_genre = movies_cleaned.groupby(['genre_1', 'genre_2'])

# Calculate the mean gross for each genre combination, sort, and get the top 5
PopGenre = movies_by_genre['gross'].mean().sort_values(ascending=False).head(5).reset_index()

# Rename columns for clarity
PopGenre.columns = ['genre_1', 'genre_2', 'mean_gross']
print(PopGenre) # Changed 'PopGenere' to 'PopGenre'

     genre_1                                genre_2    mean_gross
0     Family                                 Sci-Fi  4.349495e+08
1  Adventure         Animation|Drama|Family|Musical  4.227838e+08
2  Adventure  Animation|Comedy|Drama|Family|Fantasy  3.564544e+08
3     Action   Biography|Drama|History|Thriller|War  3.501236e+08
4     Action               Adventure|Fantasy|Sci-Fi  2.966848e+08


"Checkpoint 5: Well, as it turns out. Family + Sci-Fi is the most popular combo of genres out there.

Answer : Yes Family+ Sci-Fi is the most popular combo of genres.

Subtask 3.7: Find the critic-favorite and audience-favorite actors

1: Create three new dataframes namely, Meryl_Streep, Leo_Caprio, and Brad_Pitt which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the actor_1_name column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.

2:Append the rows of all these dataframes and store them in a new dataframe named Combined.

3: Group the combined dataframe using the actor_1_name column.

4: Find the mean of the num_critic_for_reviews and num_user_for_review and identify the actors which have the highest mean.1:

In [104]:
[] # Write your code for creating three new dataframes here

Meryl_Streep = movies_cleaned[movies_cleaned['actor_1_name'] == 'Meryl Streep']
Leo_Caprio = movies_cleaned[movies_cleaned['actor_1_name'] == 'Leonardo DiCaprio']
Brad_Pitt = movies_cleaned[movies_cleaned['actor_1_name'] == 'Brad Pitt']

In [105]:
 # Include all movies in which Meryl_Streep is the lead"
 # Include all movies in which Meryl_Streep is the lead
Meryl_Streep = movies_cleaned[movies_cleaned['actor_1_name'] == 'Meryl Streep']
print(Meryl_Streep['movie_title'])  # Access the 'movie_title' column using bracket notation

410             It's Complicated 
1106              The River Wild 
1204               Julie & Julia 
1408       The Devil Wears Prada 
1483             Lions for Lambs 
1575               Out of Africa 
1618                Hope Springs 
1674              One True Thing 
1925                   The Hours 
2781               The Iron Lady 
3135    A Prairie Home Companion 
Name: movie_title, dtype: object


In [106]:
# Include all movies in which Leo_Caprio is the lead
Leo_Caprio = movies_cleaned[movies_cleaned['actor_1_name'] == 'Leonardo DiCaprio']
print(Leo_Caprio ['movie_title'])  # Access the 'movie_title' column using bracket notation

26                       Titanic 
50              The Great Gatsby 
97                     Inception 
179                 The Revenant 
257                  The Aviator 
296             Django Unchained 
307                Blood Diamond 
308      The Wolf of Wall Street 
326            Gangs of New York 
361                 The Departed 
452               Shutter Island 
641                 Body of Lies 
911          Catch Me If You Can 
990                    The Beach 
1114          Revolutionary Road 
1422    The Man in the Iron Mask 
1453                    J. Edgar 
1560      The Quick and the Dead 
2067               Marvin's Room 
2757              Romeo + Juliet 
3476            The Great Gatsby 
Name: movie_title, dtype: object


In [107]:
Brad_Pitt =[] # Include all movies in which Brad_Pitt is the lead
# Include all movies in which Brad_Pitt is the lead
Brad_Pitt = movies_cleaned[movies_cleaned['actor_1_name'] == 'Brad_Pitt ']
print(Brad_Pitt ['movie_title'])  # Access the 'movie_title' column using bracket notation

Series([], Name: movie_title, dtype: object)


In [108]:
# Write your code for creating three new dataframes here

Meryl_Streep = movies_cleaned[movies_cleaned['actor_1_name'] == 'Meryl Streep']
Leo_Caprio = movies_cleaned[movies_cleaned['actor_1_name'] == 'Leonardo DiCaprio']
Brad_Pitt = movies_cleaned[movies_cleaned['actor_1_name'] == 'Brad Pitt']
# Write your code for grouping the combined dataframe here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt]) #Concatenate the three dataframes
actor_group = Combined.groupby('actor_1_name')  #Group by 'actor_1_name'
print(actor_group)

In [109]:
 # Write the code for finding the mean of critic reviews and audience reviews here

 # Write the code for finding the mean of critic reviews and audience reviews here
# Write your code for creating three new dataframes here

Meryl_Streep = movies_cleaned[movies_cleaned['actor_1_name'] == 'Meryl Streep']
Leo_Caprio = movies_cleaned[movies_cleaned['actor_1_name'] == 'Leonardo DiCaprio']
Brad_Pitt = movies_cleaned[movies_cleaned['actor_1_name'] == 'Brad Pitt']

# Write your code for grouping the combined dataframe here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt]) #Concatenate the three dataframes
actor_group = Combined.groupby('actor_1_name')  #Group by 'actor_1_name'


# Calculate the mean of critic and audience reviews for each actor
critic_reviews_mean = actor_group['num_critic_for_reviews'].mean()
audience_reviews_mean = actor_group['num_user_for_reviews'].mean()

# Print the results
print("Mean Critic Reviews:")
print(critic_reviews_mean)
print("\nMean Audience Reviews:")
print(audience_reviews_mean)

Mean Critic Reviews:
actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64

Mean Audience Reviews:
actor_1_name
Brad Pitt            742.352941
Leonardo DiCaprio    914.476190
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64


Checkpoint 6: Leonardo has aced both the lists

Yes, Leonardo has aced in both the lists.